### PR TITLE
feat(bgt): BeraCrocLPToken

### PR DESCRIPTION
**Pull request by ocnc**
## Summary by CodeRabbit

- **New Features**
	- Introduced a contract for BGT token rewards, including functions for accruing, distributing, and claiming rewards.
	- Added support for Cosmos-related types and functions in smart contracts.
	- Implemented a simple ERC20 token with EIP-2612 functionality.
	- Defined an interface for managing rewards in smart contracts.
	- Created a contract for managing Bera Croc LP ERC20 tokens, including liquidity depositing and withdrawing.
	- Added a new event for Bera Croc LP token creation.
	- Introduced functionality for batch swapping of tokens in the CrocSwapDex contract.
- **Enhancements**
	- Updated smart contracts to include new import statements and instances for improved functionality.
	- Modified liquidity burning function to use a specific price threshold.
	- Enhanced error handling in the `mintOverPool` function with a new `try-catch` block.
- **Bug Fixes**
	- Fixed an issue with liquidity parameter validations in the BeraCrocLpErc20 contract.

### CODE REVIEW

File: contracts/periphery/BeraCrocLpErc20.sol [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    uint256 public immutable poolType;
+    
+    constructor (address base, address quote, uint256 poolIdx)
+        ERC20 ("Bera Croc LP ERC20 Token", "LP-BeraCroc", 18) {
+
+        // CrocSwap protocol uses 0x0 for native ETH, so it's possible that base
+        // token could be 0x0, which means the pair is against native ETH. quote
+        // will never be 0x0 because native ETH will always be the base side of
+        // the pair.
+        require(quote != address(0) && quote > base, "Invalid Token Pair");
```
Comment (po-bera) [inline contracts/periphery/BeraCrocLpErc20.sol]:
```suggestion
        require(quote > base, "Invalid Token Pair");
```

File: contracts/mixins/MarketSequencer.sol [hunk 2]
```diff
@@ -185,11 +186,17 @@ contract MarketSequencer is TradeMatcher {
     function mintOverPool (uint128 liq, PoolSpecs.PoolCursor memory pool,
                            uint128 minPrice, uint128 maxPrice, address lpConduit)
         internal returns (int128 baseFlow, int128 quoteFlow) {
-        CurveMath.CurveState memory curve = snapCurveInRange
+        // Attempt to treat lpConduit as a BeraCrocLpErc20 contract and call a method
+        try BeraCrocLpErc20(lpConduit).getPendingBGT(address(this)) {
```
Comment (po-bera) [inline contracts/mixins/MarketSequencer.sol]:
```suggestion
        try BeraCrocLpErc20(lpConduit).previewAccruedBGT(address(this)) {
```

File: contracts/libraries/CosmosTypes.sol [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * @dev This contract uses types in the Cosmos library.
+ */
+contract CosmosTypes {
+    function coin(Cosmos.Coin calldata) public pure {}
+
+    function pageRequest(Cosmos.PageRequest calldata) public pure {}
+
+    function pageResponse(Cosmos.PageResponse calldata) public pure {}
+
+    function codecAny(Cosmos.CodecAny calldata) public pure {}
```
[the remark this answers is not in the document]
Comment (rustybera) [reply contracts/libraries/CosmosTypes.sol]: I have the same question.

File: contracts/libraries/ERC20.sol
Comment (rustybera) [inline contracts/libraries/ERC20.sol:1]: I advise against copying solady's source. The [node module](https://www.npmjs.com/package/solady) should be used instead.

File: contracts/libraries/IRewardsModule.sol
Comment (rustybera) [inline contracts/libraries/IRewardsModule.sol:1]: Move it to `interfaces`?

File: contracts/periphery/BeraCrocLpErc20.sol [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    function withdrawCrocLiq (address sender, bytes32 pool,
+                              int24 lowerTick, int24 upperTick, uint128 seeds,
+                              uint72) public override returns (bool) {
+        require(pool == poolHash, "Wrong pool");
+        require(lowerTick == 0 && upperTick == 0, "Non-BeraCroc LP Deposit");
+        _burn(sender, seeds);
+        return true;
+    }
+}
```
Comment (rustybera) [inline contracts/periphery/BeraCrocLpErc20.sol]: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

**Review by rustybera — CHANGES_REQUESTED** [review]

File: contracts/callpaths/ColdPath.sol [hunk 5]
```diff
@@ -129,6 +130,9 @@ contract ColdPath is MarketSequencer, DepositDesk, ProtocolAccount {
         
         (int128 baseFlow, int128 quoteFlow) = initCurve(pool, price, initLiq);
         settleInitFlow(lockHolder_, base, baseFlow, quote, quoteFlow);
+
+        BeraCrocLpErc20 lpToken = new BeraCrocLpErc20(base, quote, poolIdx);
```
Comment (rustybera) [inline contracts/callpaths/ColdPath.sol]: I think we should use the `CREATE2` opcode to have deterministic addresses for counterfactual deployment.
Reference:
1. https://github.com/Uniswap/v2-core/blob/master/contracts/UniswapV2Factory.sol#L28-L32
2. https://blog.openzeppelin.com/getting-the-most-out-of-create2
3. https://book.getfoundry.sh/tutorials/create2-tutorial

**Review by rustybera — CHANGES_REQUESTED** [review]

File: contracts/callpaths/ColdPath.sol (hunk 5 above)
Comment (po-bera) [reply contracts/callpaths/ColdPath.sol]: ~~If we want to deploy Crocswap to the current testnet, we cannot use CREATE2 as it is not supported.~~

Comment (rustybera) [reply contracts/callpaths/ColdPath.sol]: That's a huge surprise.

File: contracts/mixins/MarketSequencer.sol (hunk 2 above)
Comment (rustybera) [reply contracts/mixins/MarketSequencer.sol]: I'm not sure if the error will be caught if `lpConduit` doesn't exist without writing a unit test. But we may as well add a handshake function to `BeraCrocLpErc20`.

Comment (ocnc) [reply contracts/mixins/MarketSequencer.sol]: changedd to previewAccrued

File: contracts/libraries/ERC20.sol
Comment (ocnc) [reply contracts/libraries/ERC20.sol:1]: importing now

File: contracts/callpaths/ColdPath.sol [hunk 6]
```diff
@@ -131,7 +131,8 @@ contract ColdPath is MarketSequencer, DepositDesk, ProtocolAccount {
         (int128 baseFlow, int128 quoteFlow) = initCurve(pool, price, initLiq);
         settleInitFlow(lockHolder_, base, baseFlow, quote, quoteFlow);
 
-        BeraCrocLpErc20 lpToken = new BeraCrocLpErc20(base, quote, poolIdx);
+        bytes32 salt = keccak256(abi.encodePacked(base, quote));
+        BeraCrocLpErc20 lpToken = new BeraCrocLpErc20{salt: salt}(base, quote, poolIdx);
```
Comment (rustybera) [inline contracts/callpaths/ColdPath.sol]: Actually constructor arguments change the init code hash. In order to make the computation of LP address easier and deduplicate, we have to use Uniswap's pattern to init the pool.
https://github.com/Uniswap/v2-core/blob/master/contracts/UniswapV2Factory.sol#L33

**Review by rustybera — CHANGES_REQUESTED** [review]

File: contracts/callpaths/ColdPath.sol (hunk 6 above)
Comment (ocnc) [reply contracts/callpaths/ColdPath.sol]: applied changes

File: contracts/periphery/BeraCrocLpErc20.sol [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        return true;
+    }
+
+        // called once by the factory at time of deployment
+    function initialize(address _base, address _quote, uint256 _idx) external {
+        // CrocSwap protocol uses 0x0 for native ETH, so it's possible that base
+        // token could be 0x0, which means the pair is against native ETH. quote
+        // will never be 0x0 because native ETH will always be the base side of
+        // the pair.
+        require(_quote != address(0) && _base != _quote && _quote > _base, "Invalid Token Pair");
```
Comment (po-bera) [inline contracts/periphery/BeraCrocLpErc20.sol]:
```suggestion
        require(_quote > _base, "Invalid Token Pair");
```

File: contracts/periphery/BeraCrocLpErc20.sol [hunk 8]
```diff
@@ -8,24 +8,14 @@ import "../interfaces/ICrocLpConduit.sol";
 
 contract BeraCrocLpErc20 is ICrocLpConduit, BGTEligibleERC20 {
 
-    bytes32 public immutable poolHash;
-    address public immutable baseToken;
-    address public immutable quoteToken;
-    uint256 public immutable poolType;
+    address public factory;
```
Comment (rustybera) [inline contracts/periphery/BeraCrocLpErc20.sol]:
```suggestion
    address public immutable factory;
```

File: contracts/periphery/BeraCrocLpErc20.sol [hunk 9]
```diff
@@ -45,4 +35,18 @@ contract BeraCrocLpErc20 is ICrocLpConduit, BGTEligibleERC20 {
...
         return true;
     }
+
+        // called once by the factory at time of deployment
+    function initialize(address _base, address _quote, uint256 _idx) external {
+        // CrocSwap protocol uses 0x0 for native ETH, so it's possible that base
+        // token could be 0x0, which means the pair is against native ETH. quote
+        // will never be 0x0 because native ETH will always be the base side of
+        // the pair.
+        require(_quote != address(0) && _base != _quote && _quote > _base, "Invalid Token Pair");
```
Comment (rustybera) [inline contracts/periphery/BeraCrocLpErc20.sol]:
```suggestion
        require(_base != address(0) && _quote > _base, "Invalid Token Pair");
```

File: contracts/periphery/BeraCrocLpErc20.sol (hunk 7 above)
Comment (ocnc) [reply contracts/periphery/BeraCrocLpErc20.sol]: oops, accidentally reintroduced that

**Review by rustybera — APPROVED** [review]

File: contracts/periphery/BeraCrocLpErc20.sol (hunk 9 above)
Comment (ocnc) [reply contracts/periphery/BeraCrocLpErc20.sol]: similar to @po-bera 's suggestion. simply checking quote > base should be sufficient

File: contracts/periphery/BeraCrocLpErc20.sol (hunk 8 above)
Comment (ocnc) [reply contracts/periphery/BeraCrocLpErc20.sol]: done

File: contracts/periphery/BeraCrocLpErc20.sol (hunk 9 above)
Comment (rustybera) [reply contracts/periphery/BeraCrocLpErc20.sol]: `quote == 1` and `base == 0` would pass.

Comment (po-bera) [reply contracts/periphery/BeraCrocLpErc20.sol]: We still want `base == 0` for `BERA`?

Comment (rustybera) [reply contracts/periphery/BeraCrocLpErc20.sol]: I see.

**Pull request merged**